x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datatiers/NodesDataTiersUsageTransportAction.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.datatiers;
+
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.FailedNodeException;
+import org.elasticsearch.action.admin.indices.stats.CommonStatsFlags;
+import org.elasticsearch.action.admin.indices.stats.IndexShardStats;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.nodes.BaseNodesRequest;
+import org.elasticsearch.action.support.nodes.BaseNodesResponse;
+import org.elasticsearch.action.support.nodes.TransportNodesAction;
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.elasticsearch.cluster.routing.allocation.DataTier;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.store.StoreStats;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.indices.NodeIndicesStats;
+import org.elasticsearch.tasks.CancellableTask;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * Sources locally data tier usage stats mainly indices and shard sizes grouped by preferred data tier.
+ */
+public class NodesDataTiersUsageTransportAction extends TransportNodesAction<
```

Comment:
Thanks for working on this Mary. 

I'm probably missing something, but what's the difference between this new transport action and potentially continuing to use the node stats action with an empty set of metrics but with the indices flags configured? 

i.e. 
```
  NodesStatsRequest statsRequest = new NodesStatsRequest(nodes);
                statsRequest.clear()
                    .indices(new CommonStatsFlags().clear().set(CommonStatsFlags.Flag.Docs, true).set(CommonStatsFlags.Flag.Store, true));
```

Reply:
I was trying to make the transport action as lean as possible. I thought this was achieving that but a benchmark would definitely help figure out if it's worth it or not.

Reply:
I am going to list here samples of alternative approaches:

- Using a trimmed version of node stats,  we need only store & docs but we do need them on shard level otherwise they are aggregated per node:
```
GET _nodes/stats/indices/store,docs?level=shards
{
  "_nodes": {
    "total": 1,
    "successful": 1,
    "failed": 0
  },
  "cluster_name": "elasticsearch",
  "nodes": {
    "Ypq3Z2f_S96Rqj_UMPe2Kg": {
      "timestamp": 1699356609136,
      "name": "Marias-MBP-2.home",
      "transport_address": "127.0.0.1:9300",
      "host": "127.0.0.1",
      "ip": "127.0.0.1:9300",
      "roles": [...],
      "attributes": {...},
      "indices": {
        "docs": {
          "count": 110,
          "deleted": 38
        },
        "store": {
          "size_in_bytes": 2826327,
          "total_data_set_size_in_bytes": 2826327,
          "reserved_in_bytes": 0
        },
        "shards": {
          ".internal.alerts-security.alerts-default-000001": [
            {
              "0": {
                "routing": {
                  "state": "STARTED",
                  "primary": true,
                  "node": "Ypq3Z2f_S96Rqj_UMPe2Kg",
                  "relocating_node": null
                },
                "docs": {
                  "count": 0,
                  "deleted": 0
                },
                "store": {
                  "size_in_bytes": 249,
                  "total_data_set_size_in_bytes": 249,
                  "reserved_in_bytes": 0
                },
                "commit": {
                  "id": "C32+lLS09h+Ak5NNXF+VMQ==",
                  "generation": 3,
                  "user_data": {
                    "local_checkpoint": "-1",
                    "es_version": "8500003",
                    "min_retained_seq_no": "0",
                    "max_seq_no": "-1",
                    "history_uuid": "mwJ-o4Q3QMu5qW0CbQ4HBg",
                    "max_unsafe_auto_id_timestamp": "-1",
                    "translog_uuid": "mUH7cQcaQQmhHOGbhJVOtg"
                  },
                  "num_docs": 0
                },
                "seq_no": {
                  "max_seq_no": -1,
                  "local_checkpoint": -1,
                  "global_checkpoint": -1
                },
                "retention_leases": {
                  "primary_term": 1,
                  "version": 1,
                  "leases": [
                    {
                      "id": "peer_recovery/Ypq3Z2f_S96Rqj_UMPe2Kg",
                      "retaining_seq_no": 0,
                      "timestamp": 1699356264934,
                      "source": "peer recovery"
                    }
                  ]
                },
                "shard_path": {
                  "state_path": "/Users/gmarouli/projects/elasticsearch/build/distribution/local/elasticsearch-8.12.0-SNAPSHOT/data",
                  "data_path": "/Users/gmarouli/projects/elasticsearch/build/distribution/local/elasticsearch-8.12.0-SNAPSHOT/data",
                  "is_custom_data_path": false
                },
                "search_idle": true,
                "search_idle_time": 346455
              }
            }
          ],....
```
This response contains the information we want but for every shard the footprint is quite big. Benchmark will show exactly how much.

- Using the indices stats is a bit more compact response but not too much and I was concerned that we might encounter complications because the privileges differ between the usage action and this one:
```
GET _stats/docs,store?expand_wildcards=all
{
  "_shards": {
    "total": 18,
    "successful": 18,
    "failed": 0
  },
  "_all": {
    "primaries": {
      "docs": {
        "count": 110,
        "deleted": 5
      },
      "store": {
        "size_in_bytes": 2743520,
        "total_data_set_size_in_bytes": 2743520,
        "reserved_in_bytes": 0
      }
    },
    "total": {
      "docs": {
        "count": 110,
        "deleted": 5
      },
      "store": {
        "size_in_bytes": 2743520,
        "total_data_set_size_in_bytes": 2743520,
        "reserved_in_bytes": 0
      }
    }
  },
  "indices": {
    ".kibana_8.9.1_001": {
      "uuid": "aGZzdSIGRD-rG7d8e9guVw",
      "health": "green",
      "status": "open",
      "primaries": {
        "docs": {
          "count": 51,
          "deleted": 0
        },
        "store": {
          "size_in_bytes": 93243,
          "total_data_set_size_in_bytes": 93243,
          "reserved_in_bytes": 0
        }
      },
      "total": {
        "docs": {
          "count": 51,
          "deleted": 0
        },
        "store": {
          "size_in_bytes": 93243,
          "total_data_set_size_in_bytes": 93243,
          "reserved_in_bytes": 0
        }
      },
      "shards": {
        "0": [
          {
            "routing": {
              "state": "STARTED",
              "primary": true,
              "node": "Ypq3Z2f_S96Rqj_UMPe2Kg",
              "relocating_node": null
            },
            "docs": {
              "count": 51,
              "deleted": 0
            },
            "store": {
              "size_in_bytes": 93243,
              "total_data_set_size_in_bytes": 93243,
              "reserved_in_bytes": 0
            },
            "commit": {
              "id": "C32+lLS09h+Ak5NNXF+cfQ==",
              "generation": 13,
              "user_data": {
                "local_checkpoint": "121",
                "es_version": "8500003",
                "min_retained_seq_no": "117",
                "max_seq_no": "121",
                "history_uuid": "fY2gWHtJTGi2mUcY_gnHYg",
                "max_unsafe_auto_id_timestamp": "-1",
                "translog_uuid": "eVTs1OgeTpSKsXt5cf6HSw"
              },
              "num_docs": 51
            },
            "seq_no": {
              "max_seq_no": 125,
              "local_checkpoint": 125,
              "global_checkpoint": 125
            },
            "retention_leases": {
              "primary_term": 1,
              "version": 19,
              "leases": [
                {
                  "id": "peer_recovery/Ypq3Z2f_S96Rqj_UMPe2Kg",
                  "retaining_seq_no": 126,
                  "timestamp": 1699357291431,
                  "source": "peer recovery"
                }
              ]
            },
            "shard_path": {
              "state_path": "/Users/gmarouli/projects/elasticsearch/build/distribution/local/elasticsearch-8.12.0-SNAPSHOT/data",
              "data_path": "/Users/gmarouli/projects/elasticsearch/build/distribution/local/elasticsearch-8.12.0-SNAPSHOT/data",
              "is_custom_data_path": false
            },
            "search_idle": false,
            "search_idle_time": 11957
          }
        ]
      }
    },..
``` 

Finally, the new action passes objects like this for every tier - node combination, the main gains are that replicas do not add to the size of the response. The size of the response is influenced only. by the primary shards per node because they influence the indices and the primary shard sizes. We cannot merge these two stats because the master node will calculate certain stats over all primary shards:
```
{
   "data_hot": {
       "indices": ["index-1",....]
       "primaryShardSizes": [93243, ...]
       "totalShardCount": X,
       "totalShardSize": Y
   }
}
```
Please feel free to share any initial thoughts. I am going to work on benchmarking now, but it might take a bit because I haven't done this before.

Reply:
Ah thanks for publishing the outputs here Mary. This is super helpful.

I think we don't need to do any benchmarking as the reduced node stats response does already contain all the indices names on the node (and the extra stuff we don't need). 

I think it's fine to proceed with the new transport action you proposed in this PR.